File: internal/issue.go
package pie

import "time"

type SourceSystemType int

const (
	GitHub SourceSystemType = iota
	Jira
)

type IssueEventType int

const (
	IssueEventCreated IssueEventType = iota
	IssueEventUpdated
	IssueEventClosed
	IssueEventCommentAdded
	IssueEventReopened
	IssueEventAssigned
	IssueEventResolved
)

type Project struct {
	// SourceSystem is the source system for a project.
	SourceSystem SourceSystemType
	// Owner is the organistion or domain for a project.
	Owner string
	// Name is the project name for a project
	Name string
}

type IssueComment struct {
	// Project is the project for the issue comment.
	Project Project
	// IssueID is the ID of the issue.
	IssueID int64
	// ID is the numberical id for the comment.
	ID int64
	// Username is the name of the user that posted the comment.
	Username string
	// Body is the body of a comment.
	Body string
	// CreatedAt is the time a comment was created.
	CreatedAt time.Time
}

type Issue struct {
	// Project is the project for the issue.
	Project Project
	// ID is the numerical id for the issue.
	ID int64
	// ParentID is the parent id for the issue, which is 0 when the issue has no parent.
	ParentID int64
	// Assignee is the username for the assignee the issue.
	AssigneeUsername string
	// Title is the title for an issue.
	Title string
	// Description is the description for an issue.
	Description string
	// Labels is an array of labels for an issue.
	Labels []string
	// CreatedAt is the time an issue was created.
	CreatedAt time.Time
}

type IssueEvent struct {
	// Project is the project for the issue.
	Project Project
	// ID is the numerical id for the issue.
	ID int64
	// ParentID is the id for the parent of an issue, which is 0 when the issue has no parent.
	ParentID int64
	// Type is the type of the IssueEvent.
	Type IssueEventType
	// Assignee is the username for the assignee for the issue.
	AssigneeUsername string
	// Timestamp is the timestamp for the issue event.
	Timestamp time.Time
}

File: pie/issue.py
import datetime
from enum import IntEnum
from typing import NamedTuple


class SourceSystemType(IntEnum):
    """
    Source system type.
    """
    GITHUB = 0
    JIRA = 1


class IssueEventType(IntEnum):
    """
    Issue event type.
    """
    CREATED = 0
    UPDATED = 1
    CLOSED = 2
    COMMENT_ADDED = 3
    REOPENED = 4
    ASSIGNED = 5
    RESOLVED = 6


class Project(NamedTuple):
    """
    Represents a project.
    """
    # The source system for a project.
    source_system: SourceSystemType
    # The organisation or domain for a project.
    owner: str
    # The project name for a project.
    name: str


class IssueComment(NamedTuple):
    """
    Represents an issue comment.
    """
    # The project for the issue comment.
    project: Project
    # The ID of the issue.
    issue_id: int
    # The numerical id for the comment.
    id: int
    # The name of the user that posted the comment.
    username: str
    # The body of a comment.
    body: str
    # The time a comment was created.
    created_at: datetime.datetime


class Issue(NamedTuple):
    """
    Represents an issue.
    """
    # The project for the issue.
    project: Project
    # The numerical id for the issue.
    id: int
    # The parent id for the issue, which is 0 when the issue has no parent.
    parent_id: int
    # The username for the assignee of the issue.
    assignee_username: str
    # The title for the issue.
    title: str
    # The description for the issue.
    description: str
    # An array of labels for the issue.
    labels: list[str]
    # The time the issue was created.
    created_at: datetime.datetime


class IssueEvent(NamedTuple):
    """
    Represents an issue event.
    """
    # The project for the issue.
    project: Project
    # The numerical id for the issue.
    id: int
    # The numerical ID for a related object, such as comment.
    related_object_id: int
    # The id for the parent of an issue, which is 0 when the issue has no parent.
    parent_id: int
    # The type of the IssueEvent.
    type: IssueEventType
    # The username for the assignee for the issue.
    assignee_username: str
    # The timestamp for the issue event.
    timestamp: datetime.datetime
